GROMACS: build for target architecture based on optarch

ostueker opened the pull request:
Gromac's CMake tries persistently to detect the CPU capabilities on the build node, which makes
problems if the target platform is older than that of the build node.

This patch therefore selects appopriate GMX_SIMD cmake flags based on flags present in the optarch
string (originating from `EASYBUILD_OPTARCH`).  If no known flags are found, it falls back to cmake's auto detection of capabilities on the build node.

boegel commented on easybuild/easyblocks/g/gromacs.py:
```diff
@@ -123,6 +124,16 @@ def configure_step(self):
             # disable GUI tools
             self.cfg.update('configopts', "-DGMX_X11=OFF")
 
+            # convince to build for an older architecture then present on the build node by setting GMX_SIMD CMake flag.
```
`s/then/than`

boegel commented on easybuild/easyblocks/g/gromacs.py:
```diff
@@ -123,6 +124,16 @@ def configure_step(self):
             self.cfg.update('configopts', "-DGMX_X11=OFF")
 
+            # convince to build for an older architecture then present on the build node by setting GMX_SIMD CMake flag.
+            if build_option('optarch', default=None) and self.toolchain.comp_family() in build_option('optarch'):
```
You should also check here whether `optarch` is specified by compiler?

In theory, you could also have a hit here with `GCC` when `optarch` is just specified as a set of options (regardless of compiler), which would make this crash and burn below?

You could use something like:
```python
if isinstance(optarch, dict) and ...
```

boegel commented on easybuild/easyblocks/g/gromacs.py:
```diff
@@ -123,6 +124,16 @@ def configure_step(self):
 
+            # convince to build for an older architecture then present on the build node by setting GMX_SIMD CMake flag.
+            if build_option('optarch', default=None) and self.toolchain.comp_family() in build_option('optarch'):
+                build_arch = build_option('optarch')[self.toolchain.comp_family()]
```
try to only grab these values once, introduce variable above:

```python
optarch = build_option('optarch', default=None)
comp_fam = self.toolchain.comp_family()
```

boegel commented on easybuild/easyblocks/g/gromacs.py:
```diff
@@ -357,3 +368,25 @@ def sanity_check_step(self):
+        [2] http://www.gromacs.org/Documentation/Acceleration_and_parallelization
+        """
+        build_arch = build_arch.upper()
+        if build_arch.endswith(("AVX2")) and not LooseVersion(self.version) < LooseVersion('5.0'):
```
Does it need to *end* with `avx2`, or is having `avx2` anywhere in the list of options sufficient? If the latter, use  something like

```python
if 'AVX2' in build_arch:
```

Also, simplify the version check using positive logic (easier to understand):

```python
LooseVersion(self.version) >= LooseVersion('5.0')
```

boegel commented:
@ostueker ping on the remarks?

ostueker commented:
@boegel: Sorry for the delay but I was very busy the past two months.

I think I have addressed all of your comments in commit c8a947e.

boegel commented:
@ostueker Sorry for not getting back to this earlier...

After taking a closer look at this, I reworked this a bit in https://github.com/ostueker/easybuild-easyblocks/pull/1, please review & merge to update this PR.

Mostly style fixes & moving things around, largest change is taking into account that `optarch` maybe also be a string value (for example when using `eb --optarch=mavx2 ...`

akesandgren commented:
This should probably only be done if explicitly requested in the easyconfig by a specific option.
The reason is that this is a bad thing to do for normal builds.

Basically, always build directly on the target arch!

And the targeted gromacs simd options is not complete.

mboisson commented:
@akesandgren, that is exactly what this PR does. It only gives GROMACS a specific architecture if the easybuild optarch is specified. Otherwise it keeps doing autodetection.

boegel commented:
I was wondering about this too, but figured that deciding whether or not to enable auto-detection of architecture based on the `optarch` configuration setting makes sense, since that's exactly intended to avoid building for the host system architecture (which is the default in EasyBuild), see http://easybuild.readthedocs.io/en/latest/Controlling_compiler_optimization_flags.html.

Maybe we should disable auto-detection when the `optarch` toolchain option is set to `False` though (just like we don't use `-xHost` or `-march=native` then)...
Not sure what the alternative is then though, can we tell GROMACS *not* to target a specific architecture but to build from some common denominator?

akesandgren commented:
Isn't optarch by default true? Which would make that code kick in, or am i missing something?

boegel commented:
@akesandgren The `optarch` toolchain option is `True` by default, yes.

What is being checked here is the value for `--optarch` or `$EASYBUILD_OPTARCH`, i.e. the value for the `optarch` EasyBuild configuration option, which can be something like `mavx2` or `Intel:xCORE-AVX2;GCC:mavx2`...

akesandgren commented:
Ah, slightly confusing....

ostueker commented:
@boegel No worries, I have been crazy busy all year as well.

> After taking a closer look at this, I reworked this a bit in ostueker#1, please review & merge to update this PR.

Looks good to me, however [Gromacs 2016](http://manual.gromacs.org/documentation/2016/install-guide/index.html#simd-support) also introduced a setting `AVX_512` for Intel Haswell and newer.  
I hope to find time early in the new year to add that as well. If till PR hasn't been merged by then I'll add it, otherwise I'll just file a new one.

boegel commented:
@akesandgren We usually only have one `return` per function, at the very end. Not jumping out in the middle of a function makes it easier to parse.

There's a tiny performance loss, but it won't be measurable, we'll just be checking an empty `optarch` value and nothing will happen.

akesandgren commented:
Ok, wasn't 100% sure on how the rest of the code acts on optarch being Null.
It wasn't for performance reason :-)

boegel commented:
@akesandgren This line makes sure that `optarch` is either a proper value or an empty string if it's not defined:

```python
optarch = build_option('optarch', default='')
```

boegel commented on easybuild/easyblocks/g/gromacs.py:
```diff
@@ -70,6 +72,42 @@ def __init__(self, *args, **kwargs):
+        # default: fall back on autodetection
+        res = None
+
+        optarch = build_option('optarch', default='')
```
@ostueker This isn't correct, see https://github.com/ostueker/easybuild-easyblocks/pull/2

boegel commented on easybuild/easyblocks/g/gromacs.py:
```diff
@@ -70,6 +72,42 @@ def __init__(self, *args, **kwargs):
+        # take into account that optarch value is a dictionary if it is specified by compiler family
+        if isinstance(optarch, dict):
+            comp_fam = self.toolchain.comp_family()
+            optarch = optarch.get(comp_fam, '').upper()
```
`.upper()` is only done in case `optarch` value is a `dict`, it should always be done

also fixed in https://github.com/ostueker/easybuild-easyblocks/pull/2

boegel commented:
Looks good to go now after another quick test, thanks for your efforts on this @ostueker, and @akesandgren & @mboisson for the feedback!
